router/sleeve.go

```diff
@@ -0,0 +1,1121 @@
+		return nil, err
+	}
+
+	go fwd.run(aggChan, aggDFChan, specialChan, confirmedChan,
+		finishedChan)
+	return fwd, nil
+}
+
+func checkFeatures(features []string) error {
+	for _, f := range features {
+		if f == "sleeve" {
+			return nil
+		}
+	}
+
+	return fmt.Errorf("unsupported IntraHost features (%v)", features)
+}
+
+func (fwd *sleeveForwarder) logFrom(sender *net.UDPAddr, args ...interface{}) {
+	parts := make([]interface{}, len(args)+1)
+	parts[0] = fmt.Sprintf("->[%s|%s]:", sender, fwd.remotePeer)
+	copy(parts[1:], args)
+	log.Println(parts...)
+}
+
+func (fwd *sleeveForwarder) log(args ...interface{}) {
+	fwd.lock.RLock()
+	remoteAddr := fwd.remoteAddr
+	fwd.lock.RUnlock()
+	fwd.logFrom(remoteAddr, args...)
+}
+
+func (fwd *sleeveForwarder) SetListener(listener InterHostForwarderListener) {
+	fwd.lock.Lock()
+	fwd.listener = listener
+	fwd.lock.Unlock()
+
+	// Setting the listener confirms that the forwarder is really
+	// wanted
+	if listener != nil {
+		select {
+		case fwd.confirmedChan <- struct{}{}:
+		case <-fwd.finishedChan:
+		}
+	}
+}
+
+type curriedForward struct {
+	fwd *sleeveForwarder
+	key ForwardPacketKey
+}
+
+func (fwd *sleeveForwarder) Forward(key ForwardPacketKey) FlowOp {
+	return curriedForward{fwd, key}
+}
+
+func (f curriedForward) Send(frame []byte, dec *EthernetDecoder,
+	broadcast bool) {
+	fwd := f.fwd
+	fwd.lock.RLock()
+	haveContact := (fwd.remoteAddr != nil)
+	effectivePMTU := fwd.effectivePMTU
+	stackFrag := fwd.stackFrag
+	fwd.lock.RUnlock()
+
+	if !haveContact {
+		fwd.log("Cannot forward frame yet - awaiting contact")
+		return
+	}
+
+	srcName := f.key.SrcPeer.NameByte
+	dstName := f.key.DstPeer.NameByte
+
+	// We could use non-blocking channel sends here, i.e. drop frames
+	// on the floor when the forwarder is busy. This would allow our
+	// caller - the capturing loop in the router - to read frames more
+	// quickly when under load, i.e. we'd drop fewer frames on the
+	// floor during capture. And we could maximise CPU utilisation
+	// since we aren't stalling a thread. However, a lot of work has
+	// already been done by the time we get here. Since any packet we
+	// drop will likely get re-transmitted we end up paying that cost
+	// multiple times. So it's better to drop things at the beginning
+	// of our pipeline.
+	if dec.DF() {
+		if !frameTooBig(frame, effectivePMTU) {
+			fwd.aggregate(fwd.aggregatorDFChan, srcName, dstName,
+				frame)
+			return
+		}
+
+		// Why do we need an explicit broadcast hint here,
+		// rather than just checking the frame for a broadcast
+		// destination MAC address?  Because even
+		// non-broadcast frames can be broadcast, if the
+		// destination MAC was not in our MAC cache.
+		if broadcast {
+			log.Printf("dropping too big DF broadcast frame (%v -> %v): PMTU= %v\n", dec.IP.DstIP, dec.IP.SrcIP, effectivePMTU)
+			return
+		}
+
+		// Send an ICMP back to where the frame came from
+		fn, err := dec.makeICMPFragNeeded(effectivePMTU)
```

Comment:
Several times now I've stumbled across this code and initially thought `fn` stands for `function`. `fragNeededFrame`?